### ISSUE TITLE
trigger_recollect is not flagged to 1 in many user cases

### ISSUE THREAD

**Issue by onlinebizsoft** [magento/magento2#33459]
This is something handled in this commit https://github.com/magento/magento2/commit/08608a611f21d763aa2ddd6b759ef9abd62f88b7
But it is still lacking many user cases
- A rule with an expired date
- Edit a rule
- Product price change
- .......

### Preconditions (*)

Version 2.4.2 (I believe it happens with any version, 2.4-develop as well)

### Steps to reproduce (*)

1. Create a promotion cart rule with end_date "today"
2. Add products to cart which meet the rule to be applied the promotion rule 
3. Wait for tomorrow and visit shopping cart again

### Expected result (*)

We expect the shopping cart to be updated to remove the expired rules

### Actual result (*)

The shopping cart total is not updated

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by onlinebizsoft** [comment]
Also related to https://github.com/magento/magento2/issues/24526
@magento-engcom-team @sdzhepa @ihor-sviziev  @zhartaunik

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is reproducible:
Steps performed:
1. Admin -> Marketing -> Cart Price Rules -> Add New Rule -> Create promo cart rule by setting expiry date with today's date
2. Front-end - Sign to the website - Add a product to cart -> Apply Coupon/Discount code
3. Wait till tomorrow so that the Promo code is expired and refresh Front-end Shopping Cart page
**Issue:** Discount code is not getting removed from Shopping Cart page. However discount is not applied to the Total.
**Note:** There is no issue in Checkout page -> Coupon code is removed.
![image](https://user-images.githubusercontent.com/60198231/126115439-93fb54da-e7d1-4ef3-9328-b6e129aacda9.png)

![image](https://user-images.githubusercontent.com/60198231/126115404-9e81ca78-7d3c-4b9b-ab04-c13d7ef72731.png)

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-November 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42879` were created

**Issue Available**: @engcom-November, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by onlinebizsoft** [comment]
If you test with the mini cart (without visiting the checkout cart and checkout page) then the total will be wrong on all pages

**Comment by sdzhepa** [comment]
The current user expirience is not the best but it was done intentionally for performance and purposes.

Let's imagine that we have over 100500 created rules
and a lot of customers just add products to the cart and apply for coupons/promotions and leave the website
According to the prosed expected result, we have to update thousands of abandoned carts/quotes that will never be used.
Such behavior could produce unnecessary website load.

And also if a user goes to checkout totals will be updated

So we triaged it as P4

**Comment by onlinebizsoft** [comment]
@sdzhepa I don't think it is a problem to just update the trigger_recollect flag for thousands quote. You can make a trade off between simply changing the flag for all active quotes or only changing the flag for affected quotes but in any case, the funtionalities should be working right.
It doesn't create a performance issue when a quote need to recollect total once.

**Comment by digitalpianism** [comment]
Let me know if I should create a new ticket but I think its related to this one:

**Steps to reproduce (*)**

- Create a cart rule triggering after quantity of the same item is more than 1 for example: get 25% off the entire cart when one product qty is 2 or more.
- Add twice the same product to cart which meet the rule to be applied the promotion rule
- Rule is applied
- Update the cart qty from 2 to 1. 
- Rule is not applied anymore

**Expected result (*)**

We expect the quote item "applied_rule_ids" to be updated and not contain any rule id anymore.

**Actual result (*)**

The quote item "applied_rule_ids" still contain the inactive rule id.

**Reason**

The cause is this line in the RulesApplier.php : https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/SalesRule/Model/RulesApplier.php#L416 (which is called from this line in the validator model: https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/SalesRule/Model/Validator.php#L310)

As you can see, the RulesApplier model does a merge of the old applied rule ids and the new ones. But if you check the applyRules method, it returns all the applied rules so the merge is unnecessary.

Let me know if you need more information. 

Tested on 2.4.4-p2

**Comment by leonhelmus** [comment]
@digitalpianism Thanks for raising this concern i have resolved this by adding the following patch:

```
diff --git a/Model/Validator.php b/Model/Validator.php
--- a/Model/Validator.php
+++ b/Model/Validator.php
@@ -278,6 +278,10 @@
         if ($this->_isFirstTimeResetRun) {
             $address->setAppliedRuleIds('');
             $address->getQuote()->setAppliedRuleIds('');
+            /** @var \Magento\Quote\Model\Quote\Item $item */
+            foreach ($address->getQuote()->getItems() as $item) {
+                $item->setAppliedRuleIds('');
+            }
             $this->_isFirstTimeResetRun = false;
         }
```

**Comment by digitalpianism** [comment]
@leonhelmus this was my fix:

```
diff --git a/Magento/SalesRule/Model/Quote/Discount.php b/Magento/SalesRule/Model/Quote/Discount.php
--- a/Magento/SalesRule/Model/Quote/Discount.php
+++ b/Magento/SalesRule/Model/Quote/Discount.php
       $quote->setCartFixedRules([]);
+      $address->setAppliedRuleIds('');
+      $quote->setAppliedRuleIds('');
       foreach ($items as $item) {
-          $this->rulesApplier->setAppliedRuleIds($item, []);
+          $item->setAppliedRuleIds('');
            if ($item->getExtensionAttributes()) {
                $item->getExtensionAttributes()->setDiscounts(null);
            }
```

**Comment by leonhelmus** [comment]
@digitalpianism  Yes, that would also work, but i think the reset function would be the correct place hope it does not give other issues. Your fix will make sure it's only in that scope so thats less risky.

**Comment by digitalpianism** [comment]
@leonhelmus I have tried changing the rulesApplier class code but I was facing issues I ended up finding that the buggy line was the one I removed in my fix. Basically if I understood the code properly, the goal was to reset the applied_rule_ids on the address / quote / items but it would not work as the rulesApplier method merges the old rule ids and the one from the parameter.

**Comment by engcom-Delta** [comment]
Hi @onlinebizsoft ,

Thanks for your reporting and collaboration.
We have verified the issue in latest 2.4-develop instance and the issue is reproducible.
Kindly refer the screenshots.
**Steps to reproduce**
1. Admin -> Marketing -> Cart Price Rules -> Add New Rule -> Create promo cart rule by setting expiry date with today's date
2. Front-end - Sign to the website - Add a product to cart -> Apply Coupon/Discount code
![image](https://github.com/user-attachments/assets/1d11dd86-2aa3-4277-a470-277a5051e420)
3. Wait till tomorrow so that the Promo code is expired and refresh Front-end Shopping Cart page
Issue: Discount code is not removed from coupon file of Shopping Cart page. However discount is not applied to the Total.
![image](https://github.com/user-attachments/assets/06ced6e9-b148-49a4-a959-214a2b177141)


Thanks.

**Comment by lbajsarowicz** [comment]
Commit 08608a611f21 was a partial fix. The reporter states several cases are still uncovered: expired rules, edited rules, and price changes. Please re-QA which of these specific cases still reproduce on current 2.4-develop so the remaining work can be scoped.
